scvi/inference/posterior.py

```diff
@@ -273,125 +286,375 @@ def differential_expression_stats(self, M_sampling=100):
             selection=idx2, batchid=batchid2, n_samples=n_samples
         )
-        px_scale_mean1 = px_scale1.mean(axis=0)
-        px_scale_mean2 = px_scale2.mean(axis=0)
-        px_scale = np.concatenate((px_scale1, px_scale2), axis=0)
-        all_labels = np.concatenate(
-            (np.repeat(0, len(px_scale1)), np.repeat(1, len(px_scale2))), axis=0
-        )
-        if genes is not None:
-            px_scale = px_scale[:, self.gene_dataset.genes_to_index(genes)]
-        bayes1 = get_bayes_factors(
-            px_scale,
-            all_labels,
-            cell_idx=0,
-            M_permutation=M_permutation,
-            permutation=False,
-            sample_pairs=sample_pairs,
-        )
-        if all_stats is True:
-            bayes1_permuted = get_bayes_factors(
-                px_scale,
-                all_labels,
-                cell_idx=0,
-                M_permutation=M_permutation,
-                permutation=True,
-                sample_pairs=sample_pairs,
+
+        px_scale_mean1 = scales_batches_1["scale"].mean(axis=0)
+        px_scale_mean2 = scales_batches_2["scale"].mean(axis=0)
+
+        # Sampling pairs
+        # The objective of code section below is to ensure than the samples of normalized
+        # means we consider are conditioned on the same batch id
+        batchid1_vals = np.unique(scales_batches_1["batch"])
+        batchid2_vals = np.unique(scales_batches_2["batch"])
+        if set(batchid1_vals) == set(batchid2_vals):
+            n_batches = len(set(batchid1_vals))
+            n_samples_per_batch = (
+                M_permutation // n_batches if M_permutation is not None else None
             )
-            bayes2 = get_bayes_factors(
-                px_scale,
-                all_labels,
-                cell_idx=1,
+            scales_1 = []
+            scales_2 = []
+            for batch_val in set(batchid1_vals):
+                # Select scale samples that originate from the same batch id
+                scales_1_batch = scales_batches_1["scale"][
+                    scales_batches_1["batch"] == batch_val
+                ]
+                scales_2_batch = scales_batches_2["scale"][
+                    scales_batches_2["batch"] == batch_val
+                ]
+
+                # Create more pairs
+                scales_1_local, scales_2_local = pairs_sampler(
+                    scales_1_batch,
+                    scales_2_batch,
+                    use_permutation=use_permutation,
+                    M_permutation=n_samples_per_batch,
+                )
+                scales_1.append(scales_1_local)
+                scales_2.append(scales_2_local)
+            scales_1 = np.concatenate(scales_1, axis=0)
+            scales_2 = np.concatenate(scales_2, axis=0)
+        else:
+            # As it is, this function might work poorly when the different batch ids differ
+            warnings.warn("batchid1 != batchid2: batches are ignored")
```

Comment:
is it fair to say "batches are ignored"? Even if the sets aren't equal you still "put" them into the batches in the argument to this function. for example if batchid1 = (1, 2) and batchid2 = (3, 4), you're still comparing group 1 in batches 1 and 2 to group 2 in batches 3, 4. Say you want to test control versus condition and you have two control batches (1 and 2) and two condition batches (3 and 4). This seems to be doing what we'd want. 

I'd give the warning (also should we be using logger warning?) but maybe say exactly what happens in this case.